File: android/app/src/main/java/com/padnote/android/AgentTaskClient.java
package com.padnote.android;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Polling task adapters for direct Hermes and PadNote Bridge v1. */
final class AgentTaskClient {
    static final class HttpStatusException extends IllegalStateException {
        final int status;
        HttpStatusException(int status, String message) { super(message); this.status=status; }
    }
    static final class Submission {
        final String remoteTaskId;
        final AgentTaskStore.Status status;
        Submission(String remoteTaskId, AgentTaskStore.Status status) {
            this.remoteTaskId = remoteTaskId; this.status = status;
        }
    }

    static final class RemoteStatus {
        final AgentTaskStore.Status status;
        final String output, error, approvalId, approvalTitle, approvalDescription;
        final List<AgentTaskStore.Artifact> artifacts;
        RemoteStatus(AgentTaskStore.Status status, String output, String error,
                     String approvalId, String approvalTitle, String approvalDescription,
                     List<AgentTaskStore.Artifact> artifacts) {
            this.status=status;this.output=output;this.error=error;this.approvalId=approvalId;
            this.approvalTitle=approvalTitle;this.approvalDescription=approvalDescription;
            this.artifacts=Collections.unmodifiableList(new ArrayList<>(artifacts));
        }
    }

    private static final int MAX_JSON = 2 * 1024 * 1024;
    private final AgentHttpTransport http;
    AgentTaskClient() { this(AgentHttpTransport.production()); }
    AgentTaskClient(AgentHttpTransport http) { this.http=http; }

    Submission submit(AgentTaskStore.Task task, AgentConnectionStore.Config connection)
            throws Exception {
        requireIdentity(task,connection); requireCapability(connection,"run_submission");
        if(connection.kind==AgentConnectionStore.Kind.OPENCLAW)
            throw new IllegalStateException("当前版本尚未实现 OpenClaw 任务适配");
        Map<String,String> headers=new LinkedHashMap<>();
        headers.put("Idempotency-Key",task.clientTaskId);
        AgentHttpTransport.Response response=request("POST",runsRoot(task),connection.token,
                headers,task.submissionJson.getBytes(StandardCharsets.UTF_8),MAX_JSON);
        if(response.status!=202)throw httpError("提交任务失败",response);
        JSONObject json=new JSONObject(response.utf8());
        String key=task.transport==AgentConnectionStore.Transport.BRIDGE?"task_id":"run_id";
        String remote=required(json,key);
        if(task.transport==AgentConnectionStore.Transport.BRIDGE &&
                !task.instanceId.equals(json.optString("instance_id")))
            throw new IllegalStateException("任务响应来自其他 Agent 实例");
        return new Submission(remote,mapStatus(required(json,"status")));
    }

    RemoteStatus status(AgentTaskStore.Task task,AgentConnectionStore.Config connection)
            throws Exception {
        requireRemote(task,connection); requireCapability(connection,"run_status");
        AgentHttpTransport.Response response=request("GET",runsRoot(task)+"/"+
                AgentConnectionClient.path(task.remoteTaskId),connection.token,
                Collections.emptyMap(),null,MAX_JSON);
        if(response.status!=200)throw httpError("读取任务状态失败",response);
        JSONObject json=new JSONObject(response.utf8()); validateRemote(task,json);
        JSONObject approval=json.optJSONObject("approval");
        String approvalId="",approvalTitle="",approvalDescription="";
        if(approval!=null){
            approvalId=task.transport==AgentConnectionStore.Transport.BRIDGE
                    ?approval.optString("approval_id",""):approval.optString("request_id","");
            if(approvalId.length()>256)throw new IllegalStateException("审批请求ID过长");
            approvalTitle=plain(approval.optString("title",approval.optString("tool","需要批准")),160);
            approvalDescription=plain(approval.optString("description",approval.optString("reason",
                    approval.optString("command",""))),1000);
        }
        List<AgentTaskStore.Artifact> artifacts=new ArrayList<>();
        JSONArray values=json.optJSONArray("artifacts");
        if(values!=null&&task.transport==AgentConnectionStore.Transport.BRIDGE){
            for(int i=0;i<values.length();i++){JSONObject a=values.optJSONObject(i);if(a==null)continue;
                String id=a.optString("id",""); long size=a.optLong("size_bytes",-1);
                String sha=a.optString("sha256",""); if(!id.isEmpty()&&size>=0&&size<=100L*1024*1024&&sha.matches("[a-fA-F0-9]{64}"))
                    artifacts.add(new AgentTaskStore.Artifact(id,plain(a.optString("name","artifact"),256),
                            a.optString("media_type","application/octet-stream"),size,sha.toLowerCase()));
            }
        }
        return new RemoteStatus(mapStatus(required(json,"status")),
                output(json.opt("output")),plain(json.optString("error",""),4000),
                approvalId,approvalTitle,approvalDescription,artifacts);
    }

    void stop(AgentTaskStore.Task task,AgentConnectionStore.Config connection)throws Exception{
        requireRemote(task,connection);requireCapability(connection,"run_stop");
        AgentHttpTransport.Response r=request("POST",runsRoot(task)+"/"+AgentConnectionClient.path(task.remoteTaskId)+"/stop",
                connection.token,Collections.emptyMap(),"{}".getBytes(StandardCharsets.UTF_8),MAX_JSON);
        if(r.status<200||r.status>=300)throw httpError("停止任务失败",r);
    }

    void approve(AgentTaskStore.Task task,AgentConnectionStore.Config connection,String decision)throws Exception{
        requireRemote(task,connection);requireCapability(connection,"run_approval_response");
        if(task.approvalId.isEmpty())throw new IllegalStateException("当前任务没有可绑定的审批请求");
        if(!"once".equals(decision)&&!"deny".equals(decision))throw new IllegalArgumentException("审批只支持 once 或 deny");
        JSONObject body=task.transport==AgentConnectionStore.Transport.BRIDGE
                ?new JSONObject().put("approval_id",task.approvalId).put("decision",decision)
                :new JSONObject().put("request_id",task.approvalId).put("choice",decision);
        AgentHttpTransport.Response r=request("POST",runsRoot(task)+"/"+AgentConnectionClient.path(task.remoteTaskId)+"/approval",
                connection.token,Collections.emptyMap(),body.toString().getBytes(StandardCharsets.UTF_8),MAX_JSON);
        if(r.status<200||r.status>=300)throw httpError("提交审批失败",r);
    }

    URL artifactUrl(AgentTaskStore.Task task,String artifactId)throws Exception{
        if(task.transport!=AgentConnectionStore.Transport.BRIDGE)throw new IllegalStateException("直连任务没有 Bridge 产物接口");
        return new URL(runsRoot(task)+"/"+AgentConnectionClient.path(task.remoteTaskId)+
                "/artifacts/"+AgentConnectionClient.path(artifactId));
    }

    private AgentHttpTransport.Response request(String method,String url,String token,
                                                 Map<String,String> headers,byte[] body,int max)throws Exception{
        AgentHttpTransport.Response response=http.execute(new AgentHttpTransport.Request(method,
                new URL(url),token,headers,body,max));AgentConnectionClient.rejectRedirect(response);return response;
    }
    private static String runsRoot(AgentTaskStore.Task task)throws Exception{
        String root=task.origin; if(task.transport==AgentConnectionStore.Transport.BRIDGE)
            return root+"/padnote/v1/agents/"+AgentConnectionClient.path(task.instanceId)+"/runs";
        if(root.endsWith("/v1/capabilities"))root=root.substring(0,root.length()-"/capabilities".length());
        return root.endsWith("/v1")?root+"/runs":root+"/v1/runs";
    }
    private static void requireIdentity(AgentTaskStore.Task task,AgentConnectionStore.Config c){
        if(!task.matches(c))throw new IllegalStateException("任务绑定的连接身份已变更，请新建任务后重新确认");
        if(c.token.isEmpty())throw new IllegalStateException("连接令牌不可用");
    }
    private static void requireRemote(AgentTaskStore.Task task,AgentConnectionStore.Config c){requireIdentity(task,c);if(task.remoteTaskId.isEmpty())throw new IllegalStateException("任务尚未取得远端ID");}
    private static void requireCapability(AgentConnectionStore.Config c,String value){if(!c.capabilities.contains(value))throw new IllegalStateException("Agent 未验证能力："+value);}
    private static void validateRemote(AgentTaskStore.Task task,JSONObject j){String key=task.transport==AgentConnectionStore.Transport.BRIDGE?"task_id":"run_id";if(!task.remoteTaskId.equals(j.optString(key)))throw new IllegalStateException("任务状态ID不匹配");if(task.transport==AgentConnectionStore.Transport.BRIDGE&&!task.instanceId.equals(j.optString("instance_id")))throw new IllegalStateException("任务状态来自其他Agent实例");}
    private static AgentTaskStore.Status mapStatus(String raw){switch(raw){case"completed":return AgentTaskStore.Status.COMPLETED;case"failed":return AgentTaskStore.Status.FAILED;case"cancelled":return AgentTaskStore.Status.CANCELLED;case"interrupted":return AgentTaskStore.Status.INTERRUPTED;case"waiting_for_approval":return AgentTaskStore.Status.WAITING_FOR_APPROVAL;case"stopping":return AgentTaskStore.Status.STOPPING;case"submitting":return AgentTaskStore.Status.SUBMITTING;case"running":return AgentTaskStore.Status.RUNNING;default:throw new IllegalStateException("Agent响应包含未知任务状态");}}
    private static String output(Object value){if(value==null||value==JSONObject.NULL)return"";return plain(value instanceof String?(String)value:String.valueOf(value),128*1024);}
    private static String plain(String v,int max){if(v==null)return"";String clean=v.replace('\u0000',' ').trim();return clean.length()>max?clean.substring(0,max):clean;}
    private static String required(JSONObject j,String k){String v=j.optString(k,"");if(v.isEmpty())throw new IllegalStateException("Agent响应缺少 "+k);return v;}
    private static IllegalStateException httpError(String prefix,AgentHttpTransport.Response r){String message="";try{JSONObject e=new JSONObject(r.utf8()).optJSONObject("error");if(e!=null)message=plain(e.optString("message",""),500);}catch(Exception ignored){}return new HttpStatusException(r.status,prefix+"（HTTP "+r.status+"）"+(message.isEmpty()?"":"："+message));}
}

File: ios/PadNote/Agent/AgentTaskClient.swift
import Foundation
import CryptoKit

public struct AgentTaskRemoteState: Equatable, Sendable {
    public let remoteTaskID: String
    public let status: AgentTaskStatus
    public let output: String?
    public let error: String?
    public let approval: AgentTaskApproval?
    public let artifacts: [AgentTaskArtifact]
}

public struct AgentTaskClient {
    private let connectionClient: AgentConnectionClient
    private let session: URLSession
    private let redirectDelegate: RedirectDelegate?

    public init(session: URLSession? = nil) {
        if let session {
            self.session = session
            self.redirectDelegate = nil
            self.connectionClient = AgentConnectionClient(session: session)
        } else {
            let delegate = RedirectDelegate()
            let configuration = URLSessionConfiguration.ephemeral
            configuration.httpCookieStorage = nil
            configuration.urlCache = nil
            let resolved = URLSession(configuration: configuration, delegate: delegate, delegateQueue: nil)
            self.session = resolved
            self.redirectDelegate = delegate
            self.connectionClient = AgentConnectionClient(session: resolved)
        }
    }

    public func submit(task: AgentTaskRecord, token: String) async throws -> AgentTaskRemoteState {
        let url = try endpoint(for: task, suffix: "runs")
        let body: Data
        switch task.connection.transport {
        case .direct:
            guard task.payload.bundleBase64 == nil else { throw AgentTaskError.capabilityUnavailable("task_bundle") }
            body = try JSONSerialization.data(withJSONObject: ["input": task.payload.input], options: [.sortedKeys])
        case .bridge:
            body = task.payload.canonicalData
            guard body.count <= 12 * 1024 * 1024 else { throw AgentTaskError.bundleTooLarge }
        }
        let object = try await connectionClient.requestJSON(
            url: url,
            method: "POST",
            token: token,
            body: body,
            headers: ["Idempotency-Key": task.payload.clientTaskID],
            maximumBytes: 1024 * 1024
        )
        let idKey = task.connection.transport == .bridge ? "task_id" : "run_id"
        guard let remoteID = object[idKey] as? String, !remoteID.isEmpty else { throw AgentTaskError.invalidResponse }
        return try Self.state(from: object, fallbackID: remoteID, fallbackStatus: .running, transport: task.connection.transport, expectedInstanceID: task.connection.instanceID)
    }

    public func status(task: AgentTaskRecord, token: String) async throws -> AgentTaskRemoteState {
        guard let remoteID = task.remoteTaskID else { throw AgentTaskError.invalidResponse }
        let url = try endpoint(for: task, suffix: "runs/\(remoteID)")
        let object = try await connectionClient.requestJSON(url: url, method: "GET", token: token, maximumBytes: 2 * 1024 * 1024)
        return try Self.state(from: object, fallbackID: remoteID, fallbackStatus: nil, transport: task.connection.transport, expectedInstanceID: task.connection.instanceID)
    }

    public func stop(task: AgentTaskRecord, token: String) async throws -> AgentTaskRemoteState {
        guard let remoteID = task.remoteTaskID else { throw AgentTaskError.invalidResponse }
        let url = try endpoint(for: task, suffix: "runs/\(remoteID)/stop")
        let body = try JSONSerialization.data(withJSONObject: [:])
        let object = try await connectionClient.requestJSON(url: url, method: "POST", token: token, body: body)
        return try Self.state(from: object, fallbackID: remoteID, fallbackStatus: .stopping, transport: task.connection.transport, expectedInstanceID: task.connection.instanceID)
    }

    public func approve(task: AgentTaskRecord, approvalID: String, decision: String, token: String) async throws -> AgentTaskRemoteState {
        guard decision == "once" || decision == "deny",
              !approvalID.isEmpty,
              approvalID == task.approval?.id,
              let remoteID = task.remoteTaskID else { throw AgentTaskError.invalidPayload }
        let url = try endpoint(for: task, suffix: "runs/\(remoteID)/approval")
        let payload: [String: Any]
        switch task.connection.transport {
        case .direct: payload = ["request_id": approvalID, "choice": decision]
        case .bridge: payload = ["approval_id": approvalID, "decision": decision]
        }
        let body = try JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys])
        let object = try await connectionClient.requestJSON(url: url, method: "POST", token: token, body: body)
        return try Self.state(from: object, fallbackID: remoteID, fallbackStatus: .running, transport: task.connection.transport, expectedInstanceID: task.connection.instanceID)
    }

    public func downloadArtifact(task: AgentTaskRecord, artifact: AgentTaskArtifact, token: String) async throws -> URL {
        guard task.connection.transport == .bridge,
              let remoteID = task.remoteTaskID,
              task.artifacts.contains(artifact),
              artifact.sizeBytes >= 0, artifact.sizeBytes <= 100 * 1024 * 1024,
              Self.safeIdentifier(artifact.id) else {
            throw AgentTaskError.artifactInvalid
        }
        let url = try endpoint(for: task, suffix: "runs/\(remoteID)/artifacts/\(artifact.id)")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 60
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (bytes, response) = try await session.bytes(for: request)
        guard response.url == url, let http = response as? HTTPURLResponse,
              (200..<300).contains(http.statusCode) else { throw AgentTaskError.invalidResponse }
        let safeName = Self.safeArtifactName(artifact.name)
        let destination = FileManager.default.temporaryDirectory
            .appendingPathComponent("PadNote-Agent-\(UUID().uuidString)-\(safeName)")
        FileManager.default.createFile(atPath: destination.path, contents: nil)
        do {
            let handle = try FileHandle(forWritingTo: destination)
            defer { try? handle.close() }
            var hasher = SHA256()
            var buffer = Data()
            buffer.reserveCapacity(64 * 1024)
            var count = 0
            for try await byte in bytes {
                buffer.append(byte)
                count += 1
                if count > 100 * 1024 * 1024 || count > artifact.sizeBytes { throw AgentTaskError.artifactInvalid }
                if buffer.count == 64 * 1024 {
                    try handle.write(contentsOf: buffer)
                    hasher.update(data: buffer)
                    buffer.removeAll(keepingCapacity: true)
                }
            }
            if !buffer.isEmpty {
                try handle.write(contentsOf: buffer)
                hasher.update(data: buffer)
            }
            guard count == artifact.sizeBytes,
                  hasher.finalize().map({ String(format: "%02x", $0) }).joined() == artifact.sha256.lowercased() else {
                throw AgentTaskError.artifactInvalid
            }
            return destination
        } catch {
            try? FileManager.default.removeItem(at: destination)
            throw error
        }
    }

    private func endpoint(for task: AgentTaskRecord, suffix: String) throws -> URL {
        let base = try AgentConnectionClient.validatedBaseURL(task.connection.endpoint)
        switch task.connection.transport {
        case .direct:
            let v1Base: URL
            if base.path.hasSuffix("/v1/capabilities") {
                v1Base = base.deletingLastPathComponent()
            } else if base.path.hasSuffix("/v1") {
                v1Base = base
            } else {
                v1Base = base.appending(path: "v1")
            }
            return try Self.appendingSafeSuffix(suffix, to: v1Base)
        case .bridge:
            guard let instanceID = task.connection.instanceID, Self.safeIdentifier(instanceID) else { throw AgentTaskError.connectionChanged }
            let root = base.appending(path: "padnote/v1/agents").appending(path: instanceID)
            return try Self.appendingSafeSuffix(suffix, to: root)
        }
    }

    private static func state(
        from object: [String: Any],
        fallbackID: String,
        fallbackStatus: AgentTaskStatus?,
        transport: AgentTransport,
        expectedInstanceID: String?
    ) throws -> AgentTaskRemoteState {
        let id = (object[transport == .bridge ? "task_id" : "run_id"] as? String) ?? fallbackID
        guard id == fallbackID || fallbackID.isEmpty else { throw AgentTaskError.invalidResponse }
        if transport == .bridge {
            guard let instanceID = object["instance_id"] as? String,
                  instanceID == expectedInstanceID else { throw AgentTaskError.invalidResponse }
        }
        let status: AgentTaskStatus
        if let rawStatus = object["status"] as? String {
            if rawStatus == "started" { status = .running }
            else if let parsed = AgentTaskStatus(rawValue: rawStatus) { status = parsed }
            else { throw AgentTaskError.invalidResponse }
        } else if let fallbackStatus { status = fallbackStatus }
        else { throw AgentTaskError.invalidResponse }
        let output = object["output"] as? String
        let error: String?
        if let value = object["error"] as? String { error = value }
        else if let value = object["error"] as? [String: Any] { error = value["message"] as? String }
        else { error = nil }
        let approval = parseApproval(object["approval"], transport: transport)
        let artifacts: [AgentTaskArtifact]
        if let raw = object["artifacts"], JSONSerialization.isValidJSONObject(raw),
           let data = try? JSONSerialization.data(withJSONObject: raw),
           let decoded = try? JSONDecoder().decode([AgentTaskArtifact].self, from: data) {
            artifacts = decoded
        } else { artifacts = [] }
        return AgentTaskRemoteState(remoteTaskID: id, status: status, output: output, error: error, approval: approval, artifacts: artifacts)
    }

    private static func parseApproval(_ raw: Any?, transport: AgentTransport) -> AgentTaskApproval? {
        guard let object = raw as? [String: Any] else { return nil }
        let key = transport == .direct ? "request_id" : "approval_id"
        guard let id = object[key] as? String, !id.isEmpty else { return nil }
        let title = object["title"] as? String
            ?? object["tool"] as? String
            ?? "需要批准"
        let details = [object["description"] as? String, object["reason"] as? String, object["command"] as? String]
            .compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: "\n")
        guard !details.isEmpty else { return nil }
        return AgentTaskApproval(id: id, title: title, description: details)
    }

    private static func safeArtifactName(_ value: String) -> String {
        let leaf = URL(fileURLWithPath: value).lastPathComponent
        let safe = leaf.unicodeScalars.map { scalar -> Character in
            CharacterSet.alphanumerics.contains(scalar) || ".-_ ".unicodeScalars.contains(scalar) ? Character(String(scalar)) : "_"
        }
        return String(safe).isEmpty ? "artifact" : String(safe)
    }

    private static func safeIdentifier(_ value: String) -> Bool {
        guard !value.isEmpty, value.count <= 200, value != ".", value != ".." else { return false }
        return value.unicodeScalars.allSatisfy {
            CharacterSet.alphanumerics.contains($0) || "-_.".unicodeScalars.contains($0)
        }
    }

    private static func appendingSafeSuffix(_ suffix: String, to base: URL) throws -> URL {
        let components = suffix.split(separator: "/", omittingEmptySubsequences: false).map(String.init)
        guard !components.isEmpty, components.allSatisfy(safeIdentifier) else { throw AgentTaskError.invalidResponse }
        return components.reduce(base) { $0.appendingPathComponent($1, isDirectory: false) }
    }

    private final class RedirectDelegate: NSObject, URLSessionTaskDelegate {
        func urlSession(
            _ session: URLSession,
            task: URLSessionTask,
            willPerformHTTPRedirection response: HTTPURLResponse,
            newRequest request: URLRequest,
            completionHandler: @escaping (URLRequest?) -> Void
        ) {
            completionHandler(nil)
        }
    }
}

public final class AgentTaskService: @unchecked Sendable {
    private let connectionStore: AgentConnectionStore
    private let taskStore: AgentTaskStore
    private let client: AgentTaskClient

    public init(
        connectionStore: AgentConnectionStore = AgentConnectionStore(),
        taskStore: AgentTaskStore = AgentTaskStore(),
        client: AgentTaskClient = AgentTaskClient()
    ) {
        self.connectionStore = connectionStore
        self.taskStore = taskStore
        self.client = client
    }

    @discardableResult
    public func create(connectionID: UUID, payload: AgentTaskPayload) throws -> AgentTaskRecord {
        guard let profile = connectionStore.profile(id: connectionID), profile.connected else {
            throw AgentTaskError.connectionUnavailable
        }
        guard profile.kind == .hermes else { throw AgentTaskError.capabilityUnavailable("OpenClaw") }
        guard profile.capabilities["run_submission"] == true else { throw AgentTaskError.capabilityUnavailable("run_submission") }
        if payload.bundleBase64 != nil {
            guard profile.transport == .bridge, profile.capabilities["task_bundle"] == true else {
                throw AgentTaskError.capabilityUnavailable("task_bundle")
            }
        }
        return try taskStore.create(profile: profile, payload: payload)
    }

    public func submit(id: UUID) async throws -> AgentTaskRecord {
        let task = try requireTask(id)
        let token = try credential(for: task)
        let remote = try await client.submit(task: task, token: token)
        return try apply(remote, to: task)
    }

    public func refresh(id: UUID) async throws -> AgentTaskRecord {
        let task = try requireTask(id)
        let token = try credential(for: task)
        let remote = try await client.status(task: task, token: token)
        return try apply(remote, to: task)
    }

    public func stop(id: UUID) async throws -> AgentTaskRecord {
        let task = try requireTask(id)
        let token = try credential(for: task)
        let remote = try await client.stop(task: task, token: token)
        return try apply(remote, to: task)
    }

    public func approve(id: UUID, approvalID: String, decision: String) async throws -> AgentTaskRecord {
        let task = try requireTask(id)
        let token = try credential(for: task)
        let remote = try await client.approve(task: task, approvalID: approvalID, decision: decision, token: token)
        return try apply(remote, to: task)
    }

    public func download(id: UUID, artifact: AgentTaskArtifact) async throws -> URL {
        let task = try requireTask(id)
        return try await client.downloadArtifact(task: task, artifact: artifact, token: credential(for: task))
    }

    private func requireTask(_ id: UUID) throws -> AgentTaskRecord {
        guard let task = try taskStore.task(id: id) else { throw AgentTaskError.taskNotFound }
        return task
    }

    private func credential(for task: AgentTaskRecord) throws -> String {
        guard let current = connectionStore.profile(id: task.connection.connectionID) else {
            throw AgentTaskError.connectionUnavailable
        }
        guard task.connection.stillMatches(current) else { throw AgentTaskError.connectionChanged }
        guard let token = connectionStore.token(reference: task.connection.credentialReference), !token.isEmpty else {
            throw AgentTaskError.connectionUnavailable
        }
        return token
    }

    private func apply(_ remote: AgentTaskRemoteState, to task: AgentTaskRecord) throws -> AgentTaskRecord {
        _ = try credential(for: task)
        return try taskStore.mutate(id: task.id, expectedRevision: task.recordRevision) { value in
            if let existing = value.remoteTaskID, existing != remote.remoteTaskID { throw AgentTaskError.invalidResponse }
            value.remoteTaskID = remote.remoteTaskID
            value.status = remote.status
            value.output = remote.output
            value.error = remote.error
            value.approval = remote.approval
            value.artifacts = remote.artifacts
        }
    }
}
